tools/halide_malloc_trace.h

```diff
@@ -0,0 +1,230 @@
+    } else if (base == 2) {
+        do { *numptr++ = dig2char[val & 0x1];  val >>= 1;   } while(val);
+    } else if (base < 10) {
+        do { *numptr++ = (val % base) + '0';   val /= base; } while(val);
+    } else {
+        do { *numptr++ = dig2char[val % base]; val /= base; } while(val);
+    }
+    int numdigits = numptr - numbuf;
+    --numptr;   // Point to the most significant digit
+
+    // Add a prefix to identify the base
+    switch (base) {
+        case 16: if (dst < end) *dst++='0'; if (dst < end) *dst++='x'; break;
+        case 10: break;
+        case 8:  if (dst < end) *dst++='0'; break;
+        case 2:  if (dst < end) *dst++='0'; if (dst < end) *dst++='b'; break;
+        default:
+            if (dst < end) *dst++ = 'B';
+            if (dst < end) *dst++ = dig2char[base & 0xf];
+            if (dst < end) *dst++ = '_';
+            break;
+    }
+
+    // Skip leading zeros (all but the least significant digit)
+    int i = numdigits;
+    while ((i > 1) && (*numptr == '0')) {
+        numptr--; i--;
+    }
+    // Copy the digits to dst
+    while (i > 0) {
+        if (dst < end) {
+            *dst++ = *numptr--; i--;
+        } else {
+            break;
+        }
+    }
+    if (dst <= end) *dst = '\0';
+    return dst;
+}
+
+// lw_memalign2str
+//
+// Describe the memory alignment of the val ptr value in the dst string
+// (not going past the end ptr of dst)
+//
+static char *lw_memalign2str(char *dst, const char *end, intptr_t val) {
+    intptr_t align_chk = 1024*1024;
+    while (align_chk > 0) {
+        if ((val & (align_chk-1)) == 0) {
+            char aunit = ' ';
+            if (align_chk >= 1024) {
+                align_chk >>= 10;
+                aunit = 'K';
+            }
+            if (align_chk >= 1024) {
+                align_chk >>= 10;
+                aunit = 'M';
+            }
+
+            dst = lw_strcpy(dst, end, "align:");
+            dst = lw_val2str(dst, end, align_chk, 10);
+            if (aunit != ' ') {
+                if (dst < end) *dst++ = aunit;
+            }
+            break;
+        }
+        align_chk >>= 1;
+    }
+    if (dst <= end) *dst = '\0';
+    return dst;
+}
+//---------------------------------------------------------------------------
+#endif // lw_* HL_MEMINFO
+
+void *halide_malloc_trace(void *user_context, size_t x) {
+    // Halide requires halide_malloc to allocate memory that can be
+    // read 8 bytes before the start and 8 bytes beyond the end.
+    // Additionally, we also need to align it to the natural vector
+    // width.
+    void *orig = malloc(x+(128+8));
+    if (orig == NULL) {
+        // Will result in a failed assertion and a call to halide_error
+        return NULL;
+    }
+    // Round up to next multiple of 128. Should add at least 8 bytes so we
+    // can fit the original pointer.
+    void *ptr = (void *)((((size_t)orig + 128) >> 7) << 7);
+    ((void **)ptr)[-1] = orig;
+
+#ifdef HL_MEMINFO
```

Comment:
Since this tracing functionality is only enabled via halide_enable_malloc_trace, I don't think the preprocessor guards are necessary here, which will make this simpler to use (only one step to enable it instead of two).